File: 프로그래머스/KaKao/표 편집.py
# 2021 카카오 채용연계형 인턴십 표 편집
def solution(n, k, cmd):
    answer = ["O"]*n
    table = {i: [i-1, i+1] for i in range(n)}
    table[0] = [None, 1]
    table[n-1] = [n-2, None]
    stack = []
    # print(table)

    for c in cmd:
        if len(c) > 1:
            c1, c2 = c.split(" ")

            if c1 == "U":
                for _ in range(int(c2)):
                    k = table[k][0]

            elif c1 == "D":
                for _ in range(int(c2)):
                    k = table[k][1]

        else:
            if c == "C":  # 삭제
                answer[k] = "X"
                pre, next = table[k]

                stack.append([pre, k, next])

                if next == None:
                    k = table[k][0]
                else:
                    k = table[k][1]

                if pre == None:
                    table[next][0] = None
                elif next == None:
                    table[pre][1] = None
                else:
                    table[pre][1] = next
                    table[next][0] = pre

            elif c == "Z":
                pre, cur, next = stack.pop()
                answer[cur] = "O"

                if pre == None:
                    table[next][0] = cur
                elif next == None:
                    table[pre][1] = cur
                else:
                    table[next][0] = cur
                    table[pre][1] = cur

    return "".join(answer)

File: 프로그래머스/KaKao/표 편집.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/81303
import Foundation

var stack: [Int] = []

func updateTypeU(_ X:Int, _ table: [(prev: Int, next: Int)], _ k: inout Int) {
    for _ in 0..<X {
        if k == 0 { return }
        k = table[k].prev
    }
}

func updateTypeD(_ X:Int, _ n: Int, _ table: [(prev: Int, next: Int)], _ k: inout Int) {
    for _ in 0..<X {
        if k == n-1 { return }
        k = table[k].next
    }
}

func updateTypeC(_ n:Int, table: inout [(prev: Int, next: Int)], inTable: inout [Bool], _ k: inout Int) {
    let removedPrev = table[k].prev
    let romovedNext = table[k].next
    stack.append(k)
    inTable[k] = false
    
    if removedPrev != -1 {
        table[removedPrev].next = romovedNext
    }
    
    if romovedNext != -1 {
        table[romovedNext].prev = removedPrev
    }
    
    k = romovedNext == -1 ? removedPrev : romovedNext
}

func updateTypeZ(_ n:Int, table: inout [(prev: Int, next: Int)], inTable: inout [Bool], _ k: inout Int) {
    if stack.isEmpty { return }
    
    let top = stack.removeLast()
    inTable[top] = true
    
    let restoredPrev = table[top].prev
    let restoredNext = table[top].next
    
    if restoredPrev != -1 {
        table[restoredPrev].next = top
    }
    
    if restoredNext != -1 {
        table[restoredNext].prev = top
    }
} 

func solution(_ n:Int, _ k:Int, _ cmd:[String]) -> String {
    var answer: String = ""
    var table: [(prev: Int, next: Int)] = []
    var k = k
    var inTable: [Bool] = Array(repeating: true, count: n)
    
    for i in 0..<n {
        if i == n-1 {
            table.append((i-1, -1))
        } else {
            table.append((i-1, i+1))
        }
    }
    
    
    for c in cmd {
        let splitedC = c.split(separator: " ")
        switch splitedC[0] {
            case "U":
            updateTypeU(Int(splitedC[1])!, table, &k)
            break
            case "D":
            updateTypeD(Int(splitedC[1])!, n, table, &k)
            break
            case "C":
            updateTypeC(n, table: &table, inTable: &inTable, &k)
            break
            default:
            updateTypeZ(n, table: &table, inTable: &inTable, &k)
            break
        }
    }
    
    for t in inTable {
        answer += t ? "O" : "X"
    }
    return answer
}